feat: added Riches on Juice Shop challenge

rishabhkeshan opened the pull request:
### Description

Added the first Web3 challenge as part of GSoC'23, "Riches on Juice Shop"

- Added a new user and user feedback in the feedback section
- Added a new page /juicy-nft with nft metadata and a form field.
    - If the user enters the correct keys the form is replaced with other details of the nft and outlink to opensea.
    - Added initial check for if the challenge is solved
- Added ethers library for mnemonic to private key conversion for user input.
- Added the required backend routes for form submission and challenge solved checks

### Userflow

- User goes to the feedback section on the Juice Shop app 
- He finds a weird comment "Please send me the juicy chatbot NFT in my wallet at /juicy-nft : '....''
- The user goes to the path '/juicy-nft' and finds an input box where he needs to enter the private key.
- The user can get the private key from the seedphrase either through his metamask or using a bip44 derivation code.
- As soon as user enters the correct private key, the input field changes to the metadata of the SBT NFT stored in this wallet and the challenge is marked as complete.


### Affirmation

- [ ] My code follows the [CONTRIBUTING.md](https://github.com/juice-shop/juice-shop/blob/master/CONTRIBUTING.md) guidelines

ShubhamPalriwala commented:
Hey @rishabhkeshan, could you update the PR description with the new userflow for the challenge please? would be really helpful in reviewing the story

ShubhamPalriwala reviewed (changes requested):
Hey @rishabhkeshan, thanks for the PR! 💯  The code quality looks clean wrt formatting and linting! Can you spend some time and try to remove the `any` statements and typecast the data and parameter types with their specific expectations if possible? Do let me know if you need any help with the same! This would help us make full utilisation of the benefits we get of using TS

rishabhkeshan commented on frontend/src/app/app.routing.ts:
```diff
@@ -265,3 +292,11 @@ export function token2 (...args: number[]) { // vuln-code-snippet neutral-line t
+  const mnemonicWallet = HDNodeWallet.fromPhrase(
+    'purpose betray marriage blame crunch monitor spin slide donate sport lift clutch'
+  )
+  // console.log("wallet",mnemonicWallet.privateKey);
```
Hey Shubham, this is an old code, we are now routing the same through the backend.

rishabhkeshan commented:
> Hey @rishabhkeshan, could you update the PR description with the new userflow for the challenge please? would be really helpful in reviewing the story

Hey @ShubhamPalriwala I have updated the PR description, I will just have a look and try to omit the any statements and get back

bkimminich commented:
https://github.com/juice-shop/juice-shop/actions/runs/5420395734/jobs/9879852378?pr=2029

New challenges require a so-called country mapping in https://github.com/juice-shop/juice-shop/blob/master/config/fbctf.yml - please add that to make the `custom-config-test` pass.

rishabhkeshan commented:
> https://github.com/juice-shop/juice-shop/actions/runs/5420395734/jobs/9879852378?pr=2029
> 
> New challenges require a so-called country mapping in https://github.com/juice-shop/juice-shop/blob/master/config/fbctf.yml - please add that to make the `custom-config-test` pass.

Hey @bkimminich I have added the same and have tested the application locally with `npm run test`, would request you to run the CI pipeline once.

ShubhamPalriwala reviewed (commented):
Hey @rishabhkeshan, congratulations on opening your first **web3** PR 💯  I just reviewed it and followed the user journey. I really like the challenge as it made me go through understanding a few basic web3 technical terms, I did face a few roadblocks on my expected happy path and here's my take:
1. The users need to know that we are expecting them to use BIP44 for derivation from the passphrase
2. the Juice Shop SBT page says "Polygon" network, that needs to change to "Ethereum" since we pass the coin name for private key generation in the next step so that the user knows its the Eth chain?
3. There should be a toast/error message when the input key is wrong
4. A Special Error message should be shown for specific corner cases where the user is close ie 
    a. inputs the public key, 
    b. uses a wrong coin
    c. passes the extended private key, etc
These cases are not much and handling them would really give the user a good direction to the final step
5. Can we make this challenge a 2 Star rather than a 3 Star?
6. There should also be a hint on the scoreboard if the user is unable to solve the challenge
7. The challenge description should be like a task like we have for other challenges for example, "Upload a file larger than 100 kB.", "View another user's shopping basket.", "Retrieve the content of C:\Windows\system.ini or /etc/passwd from the server.", etc

Let me know if you have any further queries on these, and we can discuss further! Just these small nits here and there, once the higher level things are gtg, then I'll take a deeper dive into the codebase! 🙌🏼 looking forward

EDIT: Please do not edit any of the coding challenges, thats why the RSN test is failing so can you please revert those files back for the CI to pass

rishabhkeshan commented:
> Hey @rishabhkeshan, congratulations on opening your first **web3** PR 💯 I just reviewed it and followed the user journey. I really like the challenge as it made me go through understanding a few basic web3 technical terms, I did face a few roadblocks on my expected happy path and here's my take:
> 
> 1. The users need to know that we are expecting them to use BIP44 for derivation from the passphrase
> 2. the Juice Shop SBT page says "Polygon" network, that needs to change to "Ethereum" since we pass the coin name for private key generation in the next step so that the user knows its the Eth chain?
> 3. There should be a toast/error message when the input key is wrong
> 4. A Special Error message should be shown for specific corner cases where the user is close ie
>    a. inputs the public key,
>    b. uses a wrong coin
>    c. passes the extended private key, etc
>    These cases are not much and handling them would really give the user a good direction to the final step
> 5. Can we make this challenge a 2 Star rather than a 3 Star?
> 6. There should also be a hint on the scoreboard if the user is unable to solve the challenge
> 7. The challenge description should be like a task like we have for other challenges for example, "Upload a file larger than 100 kB.", "View another user's shopping basket.", "Retrieve the content of C:\Windows\system.ini or /etc/passwd from the server.", etc
> 
> Let me know if you have any further queries on these, and we can discuss further! Just these small nits here and there, once the higher level things are gtg, then I'll take a deeper dive into the codebase! 🙌🏼 looking forward
> 
> EDIT: Please do not edit any of the coding challenges, thats why the RSN test is failing so can you please revert those files back for the CI to pass

Hey @ShubhamPalriwala thanks for the detailed review on the PR. I have made the changes as following:

1. BIP 44 or BIP 39 has become the standard for multi deterministic keys or wallets in the industry, I don't feel the need of adding the same anywhere at the moment, if you still think that should be the case I have added the challenge hint as "Use a BIP 39/44 wallet derivation or import the seedphrase into a fresh browser wallet e.g. Metamask"
2. I have changed the same to Ethereum though Polygon also shouldn't matter as both have the same private keys. I was using Polygon to confuse the user more and make the challenge tricky.
3. I have added an error message for the wrong inputs.
4. I have added special error messages for inputting the following, I have considered only ETH and BTC as they are the most commonly used wallets and we shouldn't be expecting users to go wrong even on this especially when we have mentioned the need:
- ETH Wallet Public Key
- ETH Wallet Address
- BTC Wallet Public Key
- BTC Wallet Private Key
- BTC Wallet Address
5. Yes I think this fits better in a 2 star challenge, have made the change.
6. Have added the following hint
Hint: "Use a BIP 39/44 wallet derivation or import the seedphrase into a fresh browser wallet e.g. Metamask"
7. Updated Challenge Description- "Take ownership of the wallet seed posted accidentally"

I have not made any changes to the coding challenges file and it is not part of the commit as you could check once, still let me know if you think that's the case.

rishabhkeshan commented:
Hey @ShubhamPalriwala @bkimminich 
I have ran rsn update and now both rsn and test passes successfully on local, I have also made all the changes suggested by Shubham, would be great if you could run the CI pipeline once

bkimminich commented:
Did you check the effect on the coding challenges that triggered the RSN? It's usually not enough to just `rsn:update` because that simply assumes all remaining differences are now intended and okay.

rishabhkeshan commented:
> Did you check the effect on the coding challenges that triggered the RSN? It's usually not enough to just `rsn:update` because that simply assumes all remaining differences are now intended and okay.

Hey @bkimminich apologies, I hadn't done that, just reverted to the previous commit locally, here are the files that triggered the RSN

<img width="1128" alt="image" src="https://github.com/juice-shop/juice-shop/assets/58903797/00f79119-8843-4628-abc9-01e95bb890a8">

rishabhkeshan commented:
1. As for the current CI checks, I have updated the lint error as above by running `lint:fix`.

2. CodeQL gives the password in config file failure, though I have added the new user in a similar manner as like the other users, so I assume that shouldn't be a problem.

3. I am not sure about the 3rd test fail, for 

>CI/CD Pipeline / test (windows-latest, 20) (pull_request)

bkimminich commented:
> > Did you check the effect on the coding challenges that triggered the RSN? It's usually not enough to just `rsn:update` because that simply assumes all remaining differences are now intended and okay.
> 
> Hey @bkimminich apologies, I hadn't done that, just reverted to the previous commit locally, here are the files that triggered the RSN
> 
> <img alt="image" width="1128" src="https://user-images.githubusercontent.com/58903797/253174985-00f79119-8843-4628-abc9-01e95bb890a8.png">
> I had a look in the file diff of my PR but couldn't find anything that could have triggered the RSN for these files.
> 
> EDIT: I had a look at the updated cache.json post the `rsn:update`
> 
> ![image](https://user-images.githubusercontent.com/58903797/253186096-0377928e-6d68-414d-a5da-6e16c1344158.png)
> 
> Could this be because of the changes in either app.module or app.routing files with the new imports added?
> 
> **EDIT 2:**
> 
> Hey @bkimminich I have identified why the RSN error occurs, with the new SeedPhraseLeak challenge there is an addition of the route path for it in `app.routing.ts` along with the component import, where as the `vuln-code-snippet end adminSectionChallenge scoreBoardChallenge` for these challenges happens in `app.routing.ts` post all the route paths are defined, please suggest what should I do for the same.

You need to add the new SeedPhrase-route to the 6 files mentioned in the precise location so that the RSN no longer fails. This is to make sure that no unexpected diffs appear for any existing coding challenge due to code changes, formatting etc. See also https://pwning.owasp-juice.shop/appendix/code-snippets.html#refactoring-safety-net

rishabhkeshan commented:
>
> You need to add the new SeedPhrase-route to the 6 files mentioned in the precise location so that the RSN no longer fails. This is to make sure that no unexpected diffs appear for any existing coding challenge due to code changes, formatting etc. See also https://pwning.owasp-juice.shop/appendix/code-snippets.html#refactoring-safety-net

Hey @bkimminich I have added the SeedPhraseLeak to all the file diffs, re ran `rsn:update` and have pushed.
Also my `.DS_Store` files were getting pushed as it wasn't part of the main `.gitignore`, I have added the same.

bkimminich commented:
We should rename the component `SeedPhraseLeakChallenge`, because it gives away the approach to the challenge within and also that it only exists to host a challenge. We don't have that with any other Angular components, they're alwaysd named after their purpose. So here, maybe just `NftComponent` or something would be better?

bkimminich commented:
Your code changes still break the RSN. The rule of thumb is this: If you have done any neutral changes, i.e. ones which should not affect existing coding challenges, the RSN cache should not change from your changes. If it does, it means that you have forgotten to reflect some of your changes in solution files. I ran your branch and went through the Score Board coding challenge:

<img width="692" alt="image" src="https://github.com/juice-shop/juice-shop/assets/3531020/5a0558b7-3277-4554-ba07-47d9f25628bb">

These lines shouldn't be part of the diff for the coding challenge. There's a whitespace difference between the solution files and the `app.routing.ts` file. Same for the Admin Section solution files.

Please revert your RSN cache commit, fix the whitespace issues, then run `npm run rsn` again to see that it will pass. Updating the RSN cache should only happen when changing actual lines that are part of a diff for any coding challenge.

bkimminich commented:
About the challenge itself, I do not think that "Take ownership of the wallet seed posted accidentally" makes it currently clear what to do actually:
<img width="1000" alt="image" src="https://github.com/juice-shop/juice-shop/assets/3531020/88450f56-78ad-4f9e-8637-3da0f0f02833">
The challenge name "Access Juicy Chatbot" is also easy to confuse with actual chatbot-related challenges.

When I am on the NFT page eventually, it is also not clear what to do, because "the first account's key" does not really mean anything:
<img width="600" alt="image" src="https://github.com/juice-shop/juice-shop/assets/3531020/4921c0a4-1f5b-4ded-b60c-786987e6604e">

rishabhkeshan commented:
Hi @bkimminich 

Thank you for your patient review. Based on your inputs, I've made some significant changes to improve the clarity of the challenge.

### Challenge Name and Description

Challenge Name:
Previous: Access Juicy Chatbot
**Updated: Juicy NFT Seed Saga**

Challenge Description:
Previous: Take ownership of the wallet seed posted accidentally
**Updated: Discover the hidden seed phrase to unlock the Juicy Chatbot NFT wallet.**

I realized that the previous description did not clearly cover the 2 parts of the challenge, finding the seedphrase and then unlocking the wallet which has the NFT, this description should make it clearer.

![image](https://github.com/juice-shop/juice-shop/assets/58903797/7f8a78f9-195f-40cd-838e-8df6f430df15)

**Note**: If we still think that this description does not highlight the fact that the seed phrase is hidden in one of the feedbacks, I have an alternate description: "Unravel the seed in a feedback to claim the Juicy Chatbot NFT."

### NFT Page Description 

Previous: Enter the first account's key to access me on Ethereum.
**Updated: Enter the private key of the first account derived from the seed phrase to access me on Ethereum.**

![image](https://github.com/juice-shop/juice-shop/assets/58903797/dcc8e537-d63e-4181-adcf-a7ed6d54fc59)


### Angular Component Name

Apologies, yes I will keep this in mind for the forthcoming challenges to not give the hints. As for the current challenge I have updated the component name to `NFTUnlockComponent`.

### RSN Update

I have reverted, fixed and pushed the RSN, there are no changes in the cache now and the test passes successfully.

I believe these changes address the issues you highlighted, making the challenge more user-friendly and coherent. Please let me know if you have any further feedback or suggestions.

rishabhkeshan commented:
Hey @bkimminich @ShubhamPalriwala I have made the following changes to the challenge as discussed:

### Challenge Detail

Name: NFT Takeover
Desc: Take over the wallet containing our official Soul Bound Token (NFT).
Hint: Find the seedphrase posted accidentally.

![Score Board](https://github.com/juice-shop/juice-shop/assets/58903797/4269f983-64cb-4c9b-ac61-b081f9447344)


### Error Data Edits

ETH Public Address Error message: Looks like you entered the public address of my ethereum wallet!
ETH Public Key Error Message: Looks like you entered the public key of my ethereum wallet!

For all other errors: Looks like you entered a non-Ethereum private key to access me.

![Screenshot 2023-07-21 at 9 05 58 PM](https://github.com/juice-shop/juice-shop/assets/58903797/d05d5758-7181-4ae5-b814-84134da5511d)


### Frontend Changes for the Input Page:

- Added Material UI input
- Removed static colors and classes and replaces with Material Ui classes.
- Changed submit button text to "Authenticate"

![Screenshot 2023-07-21 at 9 05 13 PM](https://github.com/juice-shop/juice-shop/assets/58903797/08d87a1b-3a54-4d18-99c5-1ee219366a2f)

I hope all the changes as discussed have been covered. Also, I have ran `lint`, `rsn` and `test` locally and all of them pass successfully.

ShubhamPalriwala commented on frontend/src/app/nft-unlock/nft-unlock.component.html:
```diff
@@ -0,0 +1,90 @@
+        <div>
+          <mat-card-title>Juicy Chatbot SBT</mat-card-title>
+        </div>
+        <!-- <div>Enter the private key of the first account derived from the seed phrase to access me on Ethereum.</div> -->
```
nit: can we remove this?

ShubhamPalriwala commented on frontend/src/app/nft-unlock/nft-unlock.component.html:
```diff
@@ -0,0 +1,90 @@
+          <mat-card-title>Juicy Chatbot SBT</mat-card-title>
+        </div>
+        <div class="owner-text">
+          Owned by
+          <a
+            target="_blank"
+            href="https://testnets.opensea.io/0x8343d2eb2B13A2495De435a1b15e85b98115Ce05"
+            >8343D2</a
+          >
+        </div>
+        <mat-divider class="detail-divider"></mat-divider>
+        <div class="detail-box mat-elevation-z6">
+          <div class="box-title">Account Address</div>
+          <mat-divider class="detail-divider"></mat-divider>
+          <p class="box-text">0x8343d2eb2B13A2495De435a1b15e85b98115Ce05</p>
+        </div>
+        <div class="detail-box mat-elevation-z6">
+          <div class="box-title">Description</div>
+          <mat-divider class="detail-divider"></mat-divider>
+          <p class="box-text">
+            Hurray! Find the Juice Shop SBT on
+            <a
+              target="_blank"
+              href="https://testnets.opensea.io/assets/mumbai/0xf4817631372dca68a25a18eb7a0b36d54f3dbcf7/0"
+              >Opensea here.</a
```
Can we have all of this in a single place where we store all the config data/one time variables?
what do you think @bkimminich?

ShubhamPalriwala reviewed (commented):
Hey @rishabhkeshan! The PR lgtm 👍🏼 just a couple of small nits

bkimminich commented:
🎆 It's merged!!! 🥳
